pkg/apis/apps/validation/validation.go

```diff
@@ -105,19 +105,27 @@ func ValidateStatefulSetUpdate(statefulSet, oldStatefulSet *apps.StatefulSet) fi
 	allErrs := apivalidation.ValidateObjectMetaUpdate(&statefulSet.ObjectMeta, &oldStatefulSet.ObjectMeta, field.NewPath("metadata"))
 
 	// TODO: For now we're taking the safe route and disallowing all updates to
-	// spec except for Replicas, for scaling, and Template.Spec.containers.image
+	// spec except for Replicas for scaling, and Template.Spec.Containers/InitContainers/Affinity.image
 	// for rolling-update. Enable others on a case by case basis.
 	restoreReplicas := statefulSet.Spec.Replicas
 	statefulSet.Spec.Replicas = oldStatefulSet.Spec.Replicas
 
 	restoreContainers := statefulSet.Spec.Template.Spec.Containers
 	statefulSet.Spec.Template.Spec.Containers = oldStatefulSet.Spec.Template.Spec.Containers
 
+	restoreInitContainers := statefulSet.Spec.Template.Spec.InitContainers
+	statefulSet.Spec.Template.Spec.InitContainers = oldStatefulSet.Spec.Template.Spec.InitContainers
+
+	restoreAffinity := statefulSet.Spec.Template.Spec.Affinity
+	statefulSet.Spec.Template.Spec.Affinity = oldStatefulSet.Spec.Template.Spec.Affinity
```

Comment:
This change aims to solve validation error where kubectl v1.6 does not work with server v1.5. Perhaps a better fix is to patch replicas when scaling, or create a scale endpoint for statefulset. More information can be found here: https://github.com/kubernetes/kubernetes/issues/46005